lib/ace/ext/easymotion.js

```diff
@@ -0,0 +1,281 @@
+
+    findings = {};
+
+    var matches = [];
+
+    for (var i = currentRow; i <= lastVisibleRow; i++) {
+      lang.getMatchOffsets(editor.session.getLine(i), new RegExp(lang.escapeRegExp(char), "g"))
+        .forEach(function(match) {
+          matches.push(new Range(i, match.offset, i, match.offset + match.length));
+        });
+    }
+
+    var hints = hintStrings(matches.length);
+
+    var markerData = [];
+    matches.forEach(function(range) {
+      var key = hints.shift();
+      markerData.push({key: key, range: range});
+      findings[key] = {row: range.start.row, column: range.start.column};
+    });
+
+    easymotionMarker = new EasymotionMarker(markerData);
+    editor.session.addDynamicMarker(easymotionMarker, true);
+}
+
+// input so far
+var accuInput = "";
+function easymotionJump(editor, char) {
+    accuInput = accuInput + char;
+    var idx = accuInput;
+    var keys = Object.keys(findings);
+    var matchingKeys = [];
+
+    var i;
+    for (i = 0; i < keys.length; i++) {
+      var k = keys[i];
+      if (k.slice(0, accuInput.length) === accuInput) {
+        matchingKeys.push(k);
+      }
+    }
+
+    if (matchingKeys.length === 0)
+      return true;
+    if (findings[idx] && matchingKeys.length === 1) {
+      editor.moveCursorTo(findings[idx].row, findings[idx].column);
+      editor.selection.clearSelection();
+      return true;
+    }
+
+    easymotionMarker.updateMarkers(editor);
+
+    return false;
+}
+
+var easymotionPhase = 0;
+
+function easymotionHandler(data, hashId, key, keyCode) {
+    var editor = data.editor;
+    if (keyCode == -1) return;
+    if (key.length > 1) {
+      if (key == "space") key = " ";
+      if (key[0] == "n") key = key.replace("numpad", "");
+    }
+    if ((hashId === 0 || hashId === 4) && key.length == 1)
+        return {command: "null", passEvent: true}; // wait until input event
+    if (hashId == -1)  {
+      // easymotion functionality in two phases
+      if (easymotionPhase === 0) {
+        easymotionMark(editor, key);
+        easymotionPhase++;
+        accuInput = "";
+      } else if (easymotionPhase === 1) {
+        var ret = easymotionJump(editor, key);
+        if (ret) {
+          easymotionHandlerCleanup({editor: editor});
+          easymotionPhase = 0;
+        }
+      }
+      return {command: "null"}
+    } else {
+        // some other key: exit easymotion mode
+        easymotionHandlerCleanup({editor: editor});
+        easymotionPhase = 0;
+        return {command: "null"}
+    }
+};
+
+function easymotionHandlerSetup(editor) {
+    editor.keyBinding.addKeyboardHandler(easymotionHandler);
+    editor.on("mousedown", easymotionHandlerCleanup);
+}
+
+function easymotionHandlerCleanup(event) {
+    var editor = event.editor;
+    editor.keyBinding.removeKeyboardHandler(easymotionHandler)
+    editor.off("mousedown", easymotionHandlerCleanup)
+
+    if (easymotionMarker) {
+      easymotionMarker.cleanup(editor);
```

Comment:
This doesn't work correctly if there are multiple editors. It either should store session and remove marker from it's session, or create a separate marker for each session.